### ISSUE TITLE
define global user units (mm, etc)

### ISSUE THREAD

**Issue by srio** [lucarebuffi/ShadowOui#3]
Create a new entry Shadow Tools->Preferences where some global preferences are going to be stored. 
The first is the length unit to be used. By default it is cm, but many people want mm. 
This should affect the plots that are in um!
Pay attention to the line varying gratings (check with Ruben)

**Comment by lucarebuffi** [comment]
variables to be checked:

- Grating: (Ruling Constant on X-Y plane) Ruling Density at origin [Lines/cm]

**Comment by srio** [comment]
SHADOW variables related to gratings: 
```
DIST_FAN in user_units
HOLO_R1 in user_units
HOLO_R1 in user_units
DIST_FAN in user_units
RULING: In lines*(user_units)**(-1)
RUL_A1: : In lines*(user_units)**(-2)
RUL_A2: : In lines*(user_units)**(-3)
RUL_A3: : In lines*(user_units)**(-4)
RUL_A4: : In lines*(user_units)**(-5)
```

Others: 

Attenuation coeffs in LENSES:
```
R_ATTENUATION_OBJ  in user_units**-1
R_ATTENUATION_IMA  in user_units**-1
```
Crystals:
```
THICKNESS: In cm!!!  (yes, it is only used to compute absorption in Crystals. )
R_JOHANSSON: in user_units
```
Screens/Filters: 
```
THICK:  In cm!!!  (yes, it is only used to compute absorption.)
```

It will be problems: 

In calculation of absorption for lenses using prerefl
In calculation of penetration length for mosaic crystals
In laterally graded multilayers (still not fully implemented in ShadowOui)

For that, SHADOW needs to know the units in use. So we must pass this information to SHADOW. I see two solutions: 
1) The quick-and-dirty solution: Use the variable "DUMMY" in Shadow.OE. This variable is not used and we could use to set the conversion factor from user_units to cm (or may be to SI). 
2) More elegant solution: add a new variable in SHadowOE. Problem is back compatibility, it may be  problems with old workspaces. In principle no, I tried shadow3 to ignore new variables, but certainly I will have to change ShadowVui (and I prefer not to go back to old stuff)

**Comment by lucarebuffi** [comment]
Let's use DUMMY. The Oasys widget already has the variable workspace_units_to_cm ready to be passed to SHADOW.

**Comment by RubenReininger** [comment]
On the grating.
If one wants to use user units, lets say mm, then the line density passed to SHADOW needs to be multiplied by 10. The VLS coefficients stay in user units.

**Comment by lucarebuffi** [comment]
Ok, line density will be inputed in user's unit but passed to SHADOW in lines/cm

**Comment by RubenReininger** [comment]
Yes, Sir.
Greetings

Ruben
Sent from my iPhone 631 681 4983

On Jan 19, 2016, at 03:49, Luca Rebuffi <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok, line density will be inputed in user's unit but passed to SHADOW in lines/cm

**Comment by lucarebuffi** [comment]
To Manolo: According to the decision of managing user units conversion only into shadow3, through the variable DUMMY, CompoundOpticalElements in ShadowLibExtension should be modified as well, in order to let ShadowOui pass the conversion factor to all the OE of the Compound

**Comment by lucarebuffi** [comment]
ShadowOui side of the modification is done. To close the issue a new shadow3 version is expected.

**Comment by RubenReininger** [comment]
Thanks Luca,
Let me know when it is time to lay with it
Best regards.

Ruben

On Jan 25, 2016, at 11:55 AM, Luca Rebuffi <notifications@github.com<mailto:notifications@github.com>> wrote:


ShadowOui side of the modification is done. To close the issue a new shadow3 version is expected.

**Comment by srio** [comment]
Give me two days....

**Comment by srio** [comment]
The shadow3 part is done. 

I checked it for attenuators, crystals (mosaic, laue and asymmetric bragg) gratings and lenses. There are test cases in shadow3/tests.

There are potential issues for multilayers (TO BE DONE) and may be problems for the phase in gratings (that I never use). Also, Gratings are supposed to work well for constant ruling and VLS (F_RULING =  0, 1 and 5). I do not know about holographic (2) and fan type (3). Does anybody has a test case?

**Comment by lucarebuffi** [comment]
Could we start talk about multilayers?

**Comment by srio** [comment]
I suggest to move multilayers to a separate issue. 

The roadmap is: 

- If you run the pre_mlayer preprocessor with shadow3 in command mode, and use it it in ShadowOui it should work. 

- Next step is to rewrite the multilayer preprocessor in python, using xraylib, in a similar way we did for prerefl and bragg. I already assigned an issue to me in shadow3

- Then write the widget, and test it, including the units. 

I can start looking to it, but not before next week.

**Comment by lucarebuffi** [comment]
The Pre_mlayer widget uses the method "pre_mlayer" of ShadowPreprocessorXrayLib, which uses xraylib, what's wrong with it?

**Comment by srio** [comment]
You are right. Yes!  I forgot I wrote that!!
So you can start with the widget and I will look for test cases....

**Comment by lucarebuffi** [comment]
My dear, even the widget is already done.

**Comment by srio** [comment]
My god, so it is my turn to test it!!
(I'm becoming old: I do not remember what I did!)
I'll start it asap. Today I'm presenting Oasys to my computer colleagues!

**Comment by lucarebuffi** [comment]
I hope it won't crash :D

**Comment by srio** [comment]
Probably not. But there will be issues, for sure!

**Comment by lucarebuffi** [comment]
No doubt, you are my favourite issues maker!

**Comment by RubenReininger** [comment]
Will look when I get back to the office

[image1.JPG]
Regards
Ruben
Sent from my iPhone 631 681 4983

On Feb 19, 2016, at 06:19, Luca Rebuffi <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #3<https://github.com/lucarebuffi/ShadowOui/issues/3>.
